src/stacked_borrows.rs

```diff
@@ -240,83 +271,73 @@ impl<'tcx> Stack {
+            stack.reactivate(ptr.tag, usage)?;
+            if let Some(new_bor) = new_bor {
+                stack.initiate(new_bor);
+            }
         }
+
         Ok(())
     }
 
-    pub fn memory_deallocated(
-        &mut self,
+    #[inline(always)]
+    pub fn memory_read(
+        &self,
         ptr: Pointer<Borrow>,
+        size: Size,
     ) -> EvalResult<'tcx> {
-        trace!("memory_deallocated with tag {:?}: {:?}", ptr.tag, ptr);
-        let stacks = self.stacks.get_mut();
-        for stack in stacks.iter_mut_all() {
-            // This is like mutating.
-            stack.reactivate(ptr.tag, /*force_mut*/true)?;
-        }
-        Ok(())
+        // Reads behave exactly like the first half of a reborrow-to-shr
+        self.use_and_maybe_re_borrow(ptr, size, UsageKind::Read, None)
     }
 
-    fn reborrow(
-        &self,
+    #[inline(always)]
+    pub fn memory_written(
+        &mut self,
         ptr: Pointer<Borrow>,
         size: Size,
-        new_bor: Borrow,
-        permit_redundant: bool,
     ) -> EvalResult<'tcx> {
-        let mut stacks = self.stacks.borrow_mut();
-        for stack in stacks.iter_mut(ptr.offset, size) {
-            if permit_redundant && stack.check(new_bor) {
-                // The new borrow is already active!  This can happen when creating multiple
-                // shared references from the same mutable reference.  Do nothing.
-                trace!("reborrow: New borrow {:?} is already active, not doing a thing", new_bor);
-            } else {
-                // If we are creating a uniq ref, we certainly want to unfreeze.
-                // Even if we are doing so from a raw.
-                // Notice that if this is a local, whenever we access it directly the
-                // tag here will be the bottommost `Uniq` for that local.  That `Uniq`
-                // never is accessible by the program, so it will not be used by any
-                // other access.  IOW, whenever we directly use a local this will pop
-                // everything else off the stack, invalidating all previous pointers
-                // and, in particular, *all* raw pointers.  This subsumes the explicit
-                // `reset` which the blog post [1] says to perform when accessing a local.
-                //
-                // [1] https://www.ralfj.de/blog/2018/08/07/stacked-borrows.html
-                stack.reactivate(ptr.tag, /*force_mut*/new_bor.is_uniq())?;
-                stack.initiate(new_bor)?;
-            }
-        }
+        // Writes behave exactly like the first half of a reborrow-to-mut
+        self.use_and_maybe_re_borrow(ptr, size, UsageKind::Write, None)
+    }
 
-        Ok(())
+    pub fn memory_deallocated(
+        &mut self,
+        ptr: Pointer<Borrow>,
+        size: Size,
+    ) -> EvalResult<'tcx> {
+        // This is like mutating
+        self.use_and_maybe_re_borrow(ptr, size, UsageKind::Write, None)
```

Comment:
there's nothing else happening? is the fact that the allocation is gone enough?

Reply:
I think so. What else should happen?

There's a FIXME for doing something with barriers later.

Reply:
yea... I was thinking like "clearing the borrow stack", but that's kind of of useless on a deallocated allocation... I can't think of anything here. I was mostly surprised how easy this case seems